distribution/xfer/download_test.go

```diff
@@ -270,7 +272,7 @@ func TestSuccessfulDownload(t *testing.T) {
 	layerStore := &mockLayerStore{make(map[layer.ChainID]*mockLayer)}
 	lsMap := make(map[string]layer.Store)
 	lsMap[runtime.GOOS] = layerStore
-	ldm := NewLayerDownloadManager(lsMap, maxDownloadConcurrency, func(m *LayerDownloadManager) { m.waitDuration = time.Millisecond })
+	ldm := NewLayerDownloadManager(lsMap, maxDownloadConcurrency, maxDownloadAttempts, func(m *LayerDownloadManager) { m.waitDuration = time.Millisecond })
```

Comment:
This isn't testing the new logic.  Should you add a test that makes sure that at most N attempts are made?

Reply:
Good catch! I will implement this if this right away if you guys accept my pull request (otherwise it would be a waste of hard work :) ). -1  is already tested in the validation section of the config file itself and also tested in `daemon/config/config_test.go`. A test that the max attempts is not exceeded is indeed a good one.

Reply:
I would like to propose the following tests/validations. 

I will add a validation in `daemon/config/config.go` that the value of MaxDownloadAttempts cannot be lower or equal than 0, as this would result in a instant stop of the download. The validation of < 0  is already there, so I will just extend this. 

To test if at most N attempts are made I will add a test in `distribution/xfer/download_test.go`  that will use a max download attempts of 0. This would result in an instant stop as at most 0 attempts should be made. By doing this we validate that the boundary of MaxDownloadAttempts works as we want.

Reply:
Oke, one thing I overlooked in this one is that a max download attempts of 0 will result in an 'infinite' amount of retries. Moby first tries to download and when it loses connection it does retries++ and than evaluates if retries == maxDownloadAttempts, which will therefore never evaluate 0 and will keep going forever. 

Do you guys have any suggestions on how to create this test? I need to somehow simulate an error to get into the retry loop, but how do I do this?

Reply:
Can you pass a function that returns an error, or point to pull something you know will fail?

Testing -1 should return with error, or default to once?
Testing 0 should return with error, or default to once?
Testing 1 should try twice.
Testing N should try N+1.
Testing Max should be infinite attempts?

Reply:
It's probably best to write a new test for this, and use subtests for the different values.Similar to,  for example https://github.com/moby/moby/blob/9808f6036acb12ce4b56cd8d1887eada0e4a5ed4/daemon/daemon_unix_test.go#L293-L366